src/sql/src/kafka_util.rs

```diff
@@ -40,6 +45,126 @@ enum ValType {
     Boolean,
 }
 
+impl ValType {
+    fn try_from_value(
+        &self,
+        catalog: &dyn SessionCatalog,
+        v: Option<WithOptionValue<Aug>>,
+    ) -> Result<StringOrSecret, anyhow::Error> {
+        let s = match self {
+            ValType::String { transform } => {
+                let s = String::try_from_value(v)?;
+                transform(s).into()
+            }
+            ValType::StringOrSecret => StringOrSecret::try_from_value(v)?,
+            ValType::Secret => match StringOrSecret::try_from_value(v)? {
+                StringOrSecret::String(_) => bail!("must provide a secret"),
+                secret @ StringOrSecret::Secret(_) => secret,
+            },
+            ValType::Number(lower, upper) => {
+                let i = i32::try_from_value(v)?;
+                if i < *lower || *upper < i {
+                    bail!("must be a number between {} and {}", lower, upper)
+                }
+                i.into()
+            }
+            ValType::Boolean => bool::try_from_value(v)?.into(),
+        };
+        validate_secret(catalog, &s)?;
+        Ok(s)
+    }
+}
+
+fn validate_secret(
+    catalog: &dyn SessionCatalog,
+    s: &mz_dataflow_types::connections::StringOrSecret,
+) -> Result<(), anyhow::Error> {
+    if let StringOrSecret::Secret(id) = s {
+        let item = catalog.get_item(&id);
+        if item.item_type() != CatalogItemType::Secret {
+            bail!(
+                "'{}' is not a SECRET",
+                catalog.resolve_full_name(&item.name())
+            );
+        }
+    }
+
+    Ok(())
+}
+
+trait ConfigGen {
+    fn config_key(&self) -> &str;
+    fn config_val(&self) -> ValType;
+}
```

Comment:
I'm not sure this generalizes beyond Kafka—none of the other connectors want their options in a bag of BTreeMap<String, StringOrSecret>, but instead want actual typed structs—so you might be able to just impl these directly on `KafkaConnectionOptionName`.

Reply:
This has to be a trait to impl them in this crate, and I had anticipated doing the same thing for the CSR Connector. I hear you about making these structs, though--if this is all right to get this one PR over the line, I'll adjust the subsequent PRs to move in that direction.

Reply:
Oh d'oh, they're SQL parser structs. No yeah totally fine, sorry!